<a href="https://colab.research.google.com/github/Anastasya2701/GOIT-KATRUSHENKO/blob/main/HW10_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import tensorflow as tf
from tensorflow.keras.datasets import fashion_mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
from tensorflow.keras.optimizers import Adam

In [2]:
(x_train, y_train), (x_test, y_test) = fashion_mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0
x_train = x_train.reshape(-1, 28, 28, 1)
x_test = x_test.reshape(-1, 28, 28, 1)

29515/29515 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
26421880/26421880 ━━━━━━━━━━━━━━━━━━━━ 2s 0us/step
5148/5148 ━━━━━━━━━━━━━━━━━━━━ 0s 1us/step
4422102/4422102 ━━━━━━━━━━━━━━━━━━━━ 1s 0us/step


In [6]:
model = Sequential([
    Conv2D(32, kernel_size=(3, 2), activation='relu', input_shape=(28, 28, 1)),
    MaxPooling2D(pool_size=(2, 2)),
    Dropout(0.2),

    Conv2D(64, kernel_size=(3, 3), activation='relu'),
    MaxPooling2D(pool_size=(2, 2)),
    Dropout(0.3),

    Flatten(),
    Dense(128, activation='relu'),
    Dropout(0.4),
    Dense(10, activation='softmax')
])

In [7]:
model.compile(optimizer=Adam(learning_rate=0.001),
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

In [8]:
history = model.fit(x_train, y_train, epochs=20, batch_size=64, validation_data=(x_test, y_test))

Epoch 1/20
938/938 ━━━━━━━━━━━━━━━━━━━━ 73s 76ms/step - accuracy: 0.6861 - loss: 0.8480 - val_accuracy: 0.8525 - val_loss: 0.4148
Epoch 2/20
938/938 ━━━━━━━━━━━━━━━━━━━━ 76s 69ms/step - accuracy: 0.8443 - loss: 0.4245 - val_accuracy: 0.8710 - val_loss: 0.3490
Epoch 3/20
938/938 ━━━━━━━━━━━━━━━━━━━━ 81s 68ms/step - accuracy: 0.8687 - loss: 0.3631 - val_accuracy: 0.8928 - val_loss: 0.2980
Epoch 4/20
938/938 ━━━━━━━━━━━━━━━━━━━━ 81s 68ms/step - accuracy: 0.8804 - loss: 0.3291 - val_accuracy: 0.8936 - val_loss: 0.2891
Epoch 5/20
938/938 ━━━━━━━━━━━━━━━━━━━━ 81s 66ms/step - accuracy: 0.8886 - loss: 0.3076 - val_accuracy: 0.8980 - val_loss: 0.2796
Epoch 6/20
938/938 ━━━━━━━━━━━━━━━━━━━━ 84s 68ms/step - accuracy: 0.8937 - loss: 0.2910 - val_accuracy: 0.9038 - val_loss: 0.2607
Epoch 7/20
938/938 ━━━━━━━━━━━━━━━━━━━━ 80s 67ms/step - accuracy: 0.9012 - loss: 0.2730 - val_accuracy: 0.8999 - val_loss: 0.2736
Epoch 8/20
938/938 ━━━━━━━━━━━━━━━━━━━━ 84s 69ms/step - accuracy: 0.9013 - loss: 0.2694 - 

In [9]:
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print(f"Точність на тестових даних: {test_accuracy * 100:.2f}%")

313/313 ━━━━━━━━━━━━━━━━━━━━ 4s 13ms/step - accuracy: 0.9140 - loss: 0.2468
Точність на тестових даних: 91.56%


Висновок

У цьому завданні було побудовано згорткову нейронну мережу (CNN) для класифікації зображень із датасету fashion_mnist, що дозволило досягти кращої точності в порівнянні з багатошаровою нейронною мережею (MLP). Основні переваги CNN у цьому контексті обумовлені здатністю ефективно обробляти просторову інформацію в зображеннях, що особливо важливо для класифікації зображень у різних категоріях.

Основні результати:

 1. Згорткова архітектура покращує точність завдяки використанню згорткових та пулінгових шарів, які вилучають значущі ознаки зображень і зменшують розмірність даних.
 2. Регуляризація через Dropout допомогла уникнути перенавчання, зберігаючи узагальнювальну здатність моделі.
 3. Порівняння з MLP точнсть якої була 88.96%: CNN показала вищу точність (91.56%)  на тестових даних порівняно з багатошаровою мережею, оскільки краще підходить для обробки зображень.

